fix: cap rewards to zero when price tag is explicitly set to zero

Hephaestus-V opened the pull request:
Resolves #246 

<!--
- You must link the issue number e.g. "Resolves #1234"
- Please do not replace the keyword "Resolves" https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

gentlementlegen commented:
Thank you for your first contribution.

I don't think that is the proper approach, a quick glance at the code tells me that

```typescript
return priceTagReward || Infinity;
```

Probably returns `Infinity` if `priceTagReward` is zero. Changing the `||` to `??` might already help, can your confirm?

Also, it is preferable to link a QA with your changes, either by linking a test in your repo, or showing some results.

Hephaestus-V commented:
I guess the check is necessary because `getTaskReward` returns 0 in both cases :
1) When there's an explicit "Price: 0" label 
2) When there are no price labels at all 

I assume we want our rewards to be capped at infinity when there is no price label as it will be calculated based on the interactions.
We can't just use return `priceTagReward ?? Infinity` because both cases would return 0 from `getTaskReward`.

Let me know if I misunderstood .

gentlementlegen commented:
An explicit `Price: 0 USD` is usually set for a task tagged as `Priority: 0 {Regression)` if you want to test it.

When there is no price label, by default we do not generate rewards. The behavior can be overridden by setting `requirePriceLabel` to false, and in such case we want `Infinity` indeed.

Hephaestus-V commented:
In that case `return priceTagReward || Infinity;` should work fine I guess ,
Also how do I test this if it works or not ?
I'm getting some `Export named 'HTTPParser' not found in module 'node:http'.` error while trying to run tests locally with bun .

gentlementlegen commented:
What you can do is `bun run server` that will start you a local instance. Within the `web` folder, you can compile the web UI, then when accessing `http://localhost:4000` you will have a UI allowing you to test. You can change the settings by tweaking `.ubiquity-os.config.yml`.

https://github.com/ubiquity-os-marketplace/text-conversation-rewards?tab=readme-ov-file#how-to-get-started

Hephaestus-V commented:
I couldn't find any `price: 0 ` issue in this repository and I'm not able to get result for issues in other repositories,
How do I solve this?

gentlementlegen commented:
It is okay to manually add the labels to an issue if you want to test it.

gentlementlegen commented on src/parser/processor.ts:
```diff
@@ -45,7 +45,7 @@ export class Processor {
     }
     const priceTagReward = getTaskReward(issue);
-    return priceTagReward || Infinity;
+    return priceTagReward ?? Infinity;
```
Could you create related tests?

Hephaestus-V replied:
I couldn't run existing tests using bun run test because of some `HTTPParser ` error and that's why couldn't add more tests to test this functionality but I did check it by manually changing the values in the code and using the UI and now it didn't give rewards when the `priceTagReward` is 0 i.e when the price label is 0

gentlementlegen commented:
I tested it, and it works fine, however it creates another problem: we remove users that get 0 reward from the final result comment. Since everyone is capped to `0`, it basically posts an empty message. Something should be changed, not sure if we want to include users with 0 reward back. RFC @0x4007

0x4007 commented:
I think it's a strange edge case that a funded task has 0 rewards across the board. Catch and throw a warning.

gentlementlegen commented:
@0x4007 This would happen for any task marked as "Regression", since its price tag will be `0`.

Hephaestus-V commented:
Should I implement a warning message for Price: 0 tasks instead of showing an empty comment ? 
That warning will clearly explain that no rewards were distributed because the price was explicitly set to 0.

I've to make this change in `GithubCommentModule ` if I'm not wrong .

gentlementlegen commented:
@Hephaestus-V Sounds good to me.

Hephaestus-V commented:
@gentlementlegen Can you create a test issue with `Price: 0 USD` label so that I can test if my changes work ?
Or please let me know if there is any other way I can test it ?

gentlementlegen commented:
@Hephaestus-V You can create an issue within your own repo, and manually create a tag `Price: 0 USD`, it should work.

Hephaestus-V commented:
I've extracted the warning message generation into a private method `_getZeroPriceWarningContent()`
and also refactored the code by extracting the stripped content generation into a private method `_getStrippedContent()` as I was getting cognitive complexity threshold error .

gentlementlegen commented on src/parser/github-comment-module.ts:
```diff
@@ -38,29 +38,61 @@ export class GithubCommentModule extends BaseModule {
-  async getBodyContent(data: Readonly<IssueActivity>, result: Result, stripContent = false): Promise<string> {
-    const keysToRemove: string[] = [];
+  private _getZeroPriceWarningContent(): string {
+    const warningMessage = `# Rewards Summary\n\n> [!CAUTION]\n> No rewards have been distributed for this task because it was explicitly marked with a Price: 0 label.\n\n`;
```
What you can do that's much simpler is that you can throw an error with a warning message. All you need to do is:

```typescript
throw this.context.logger.warn("No rewards have been distributed for this task because it was explicitly marked with a Price: 0 label.");
```

gentlementlegen commented:
Current QA: https://github.com/Meniole/text-conversation-rewards/issues/78#issuecomment-2758080516

I think you should just use `throw context.logger.warn("<error message");` instead. And I think that this check can occur within the pre-checks, which will save the plugin from a useless run because we know that no rewards will be generated.

gentlementlegen commented:
Latest QA: https://github.com/Meniole/text-conversation-rewards/issues/78#issuecomment-2760237968

Hephaestus-V commented:
Am I supposed to add that test?

gentlementlegen commented:
I tried running the tests and it was fine. I would advise to:
- make sure bun is up to date (1.2.6)
- delete node_modules and bun.lock
- run `bun install`
- run `bun run test`

Let me know if the issue persists.

Hephaestus-V commented:
> I tried running the tests and it was fine. I would advise to:
> 
> * make sure bun is up to date (1.2.6)
> * delete node_modules and bun.lock
> * run `bun install`
> * run `bun run test`
> 
> Let me know if the issue persists.

I tried these things but the tests still fail

gentlementlegen commented:
None of these should be required, except `LOG_LEVEL` where you can set it to `debug`. Worst case scenario the tests you will write will be executed in this pull-request so you can see the results.

Hephaestus-V commented:
I'm not able to run the tests locally they fail even after doing the changes.

Also error logs are not getting loaded on github (REF: https://github.com/ubiquity-os-marketplace/text-conversation-rewards/actions/runs/14110250036/job/39526781254 ) , how do I find out what's the error if I push the tests and then can't even see the error logs to see if it works ?

gentlementlegen commented:
For the two last errors you should be able to fix these by using the same values as the test configuration found here:
https://github.com/ubiquity-os-marketplace/text-conversation-rewards/blob/fe45efc952b3d51306d837c4ca00e06db25f8955/.github/workflows/jest-testing.yml#L19-L20

For the first one you can ignore it for now and write your tests normally. But now you are able to run the test suite so writing tests should not be a problem.

gentlementlegen commented:
Latest QA:
https://github.com/Meniole/text-conversation-rewards/issues/78#issuecomment-2767870246

Thank you for the tests!
